File: testcigar.py
import sys
import json

import jax.numpy as jnp

import cigartree
import likelihood
import h20

if len(sys.argv) != 4:
    print('Usage: {} tree.nh align.fa model.json'.format(sys.argv[0]))
    sys.exit(1)

treeFilename, alignFilename, modelFilename = sys.argv[1:]
with open(treeFilename, 'r') as f:
    treeStr = f.read()
with open(alignFilename, 'r') as f:
    alignStr = f.read()
with open(modelFilename, 'r') as f:
    modelJson = json.load (f)

ct = cigartree.makeCigarTree (treeStr, alignStr)

alphabet, mixture, indelParams = likelihood.parseHistorianParams (modelJson)
seqs, distanceToParent, parentIndex, transCounts = cigartree.getHMMSummaries (treeStr, alignStr, alphabet)

#print(transCounts)

subll = likelihood.subLogLike (seqs, distanceToParent, parentIndex, *mixture[0])
subll_total = float (jnp.sum (subll))

transMat = jnp.stack ([h20.dummyRootTransitionMatrix()] + [h20.transitionMatrix(t,indelParams,alphabetSize=len(alphabet)) for t in distanceToParent[1:]], axis=0)
transMat = jnp.log (jnp.maximum (transMat, h20.smallest_float32))
transll = transCounts * transMat
#print(transMat)
#print(transCounts)
#print(transll)
transll_total = float (jnp.sum (transll))

print (json.dumps({'loglike':{'subs':subll_total,'indels':transll_total}, 'cigartree': ct}))

File: testcigar.js

import fs from 'fs';
import { makeCigarTree, expandCigarTree, getHMMSummaries } from './cigartree.js';
import { parseHistorianParams, subLogLike, sum } from './likelihood.js';
import { transitionMatrix, dummyRootTransitionMatrix } from './h20.js';

if (process.argv.length != 5) {
    console.error('Usage: ' + process.argv[1] + ' tree.nh align.fa model.json');
    process.exit(1);
}

const [ treeFilename, alignFilename, modelFilename ] = process.argv.slice(2);
const treeStr = fs.readFileSync(treeFilename).toString();
const alignStr = fs.readFileSync(alignFilename).toString();
const modelJson = JSON.parse (fs.readFileSync(modelFilename).toString());

const ct = makeCigarTree (treeStr, alignStr);
const { leavesByColumn, internalsByColumn, branchesByColumn } = expandCigarTree (ct);

const { alphabet, mixture, indelParams } = parseHistorianParams (modelJson);
const { seqs, distanceToParent, parentIndex, transCounts } = getHMMSummaries (treeStr, alignStr, alphabet);

const { subRate, rootProb } = mixture[0];
const subll = subLogLike (seqs, distanceToParent, leavesByColumn, internalsByColumn, branchesByColumn, alphabet, subRate, rootProb);
console.warn({subll})
const subll_total = sum (subll);

let transMat = [dummyRootTransitionMatrix()].concat (distanceToParent.slice(1).map((t)=>transitionMatrix(t,indelParams,alphabet.length)));
transMat = transMat.map ((mt) => mt.map((mti) => mti.map((mtij) => Math.log (Math.max (mtij, Number.MIN_VALUE)))));
const transll = transCounts.map ((mt,t) => mt.map((mti,i) => mti.map((mtij,j) => mtij * transMat[t][i][j])));
const transll_total = sum (transll.map ((mt) => sum (mt.map(sum))));

console.log (JSON.stringify({'loglike':{'subs':subll_total,'indels':transll_total}, 'cigartree': ct}));

